src/shared/sam/cli/samCliAutoDetection.ts

```diff
@@ -0,0 +1,66 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as vscode from 'vscode'
+import * as nls from 'vscode-nls'
+import { extensionSettingsPrefix, samAboutInstallUrl } from '../../constants'
+import { DefaultSettingsConfiguration } from '../../settingsConfiguration'
+import { SamCliConfiguration } from './samCliConfiguration'
+import { DefaultSamCliLocationProvider } from './samCliLocator'
+const localize = nls.loadMessageBundle()
+
+export async function autoDetectSamCli(showMessageIfDetected: boolean): Promise<void> {
+    const samCliConfig = new SamCliConfiguration(
+        new DefaultSettingsConfiguration(extensionSettingsPrefix),
+        new DefaultSamCliLocationProvider()
+    )
+
+    await samCliConfig.initialize()
+
+    const samCliLocation = samCliConfig.getSamCliLocation()
+
+    if (!samCliLocation) {
+        notifyUserSamCliNotDetected()
+    } else if (showMessageIfDetected) {
+        vscode.window.showInformationMessage(
+            localize(
+                'AWS.samcli.autodetect.settings.updated',
+                'Settings updated. SAM CLI found at {0}',
```

Comment:
I'd considered having a state where the config didn't change (because it is already set), but then I figured a message "there is no change" was useless, and no message was misleading ("did it actually do anything?"), so this covers both cases reasonably well.

Reply:
> I figured a message "there is no change" was useless

I agree, but that's not what I'm suggesting. There was a change, just not in settings. The main scenario I'm thinking about is:

1. I attempt an action that requires SAM CLI, or I manually run the Detect SAM CLI command.
2. SAM CLI isn't installed, so I see a warning because SAM CLI can't be found.
3. I install SAM CLI.
4. I run the 'Detect SAM CLI' command again.
5. This time, it informs me that the SAM CLI *was* detected.

Reply:
Ah, I misunderstood the scenario. The setting isn't a user override, we also update the setting every time detection is run. So the only case where settings weren't updated is if the new setting is equivalent to the old setting.